File: src/org/genyris/core/Symbol.java
package org.genyris.core;

import java.net.URI;
import java.net.URISyntaxException;

import org.genyris.exception.GenyrisException;
import org.genyris.interp.Environment;
import org.genyris.interp.UnboundException;

public abstract class Symbol extends Atom implements Comparable {

    public static SimpleSymbol symbolFactory(String name, boolean escaped) {
        try {
            URI uri = new URI(name);
            if(uri.isAbsolute()) {
                return new URISymbol(name);
            }
        } catch (URISyntaxException e) { }
        return (escaped ? new EscapedSymbol(name) : new SimpleSymbol(name));
    }

    public int hashCode() { // TODO make abstract
        return getPrintName().hashCode();
    }

    public boolean equals(Object other) { // TODO make abstract
        if(!(other instanceof Symbol )){
            return false;
        }
        return getPrintName().equals(((Symbol)other).getPrintName());
    }
    public boolean isNil() {
        return false;
    }

    public abstract String getPrintName();

    public String toString() {
        return getPrintName();
    }

    public abstract Exp lookupVariableValue(Environment env) throws UnboundException;

    public abstract void setVariableValue(Environment env, Exp valu) throws UnboundException;

    public abstract void defineVariable(Environment env, Exp valu) throws GenyrisException;

    public abstract int compareTo(Object arg0);

	public Exp eval(Environment env) throws GenyrisException {
		return env.lookupVariableValue(this);
    }


}

File: src/org/genyris/core/Symbol.kt
package org.genyris.core

import org.genyris.exception.GenyrisException
import org.genyris.interp.Environment
import org.genyris.interp.UnboundException
import java.net.URI
import java.net.URISyntaxException

abstract class Symbol : Atom(), Comparable<Any?> {
    override fun hashCode(): Int { // TODO make abstract
        return this.printName!!.hashCode()
    }

    override fun equals(other: Any?): Boolean { // TODO make abstract
        if (other !is Symbol) {
            return false
        }
        return this.printName == other.printName
    }

    override fun isNil(): Boolean {
        return false
    }

    abstract val printName: String?

    override fun toString(): String {
        return this.printName!!
    }

    @kotlin.Throws(UnboundException::class)
    abstract fun lookupVariableValue(env: Environment?): Exp?

    @kotlin.Throws(UnboundException::class)
    abstract fun setVariableValue(env: Environment?, valu: Exp?)

    @kotlin.Throws(GenyrisException::class)
    abstract fun defineVariable(env: Environment?, valu: Exp?)

    abstract override fun compareTo(arg0: Any?): Int

    @kotlin.Throws(GenyrisException::class)
    override fun eval(env: Environment): Exp? {
        return env.lookupVariableValue(this)
    }


    companion object {
        fun symbolFactory(name: String, escaped: Boolean): SimpleSymbol {
            try {
                val uri = URI(name)
                if (uri.isAbsolute()) {
                    return URISymbol(name)
                }
            } catch (e: URISyntaxException) {
            }
            return (if (escaped) EscapedSymbol(name) else SimpleSymbol(name))
        }
    }

    abstract fun getPrintName(): java.lang.String
}